com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/AdvancedLaunchingConnector.java

```diff
@@ -0,0 +1,157 @@
+    public static final String HOME = "home";
+    public static final String OPTIONS = "options";
+    public static final String MAIN = "main";
+    public static final String SUSPEND = "suspend";
+    public static final String QUOTE = "quote";
+    public static final String EXEC = "vmexec";
+    private static final String CWD = "cwd";
+    private static final String ENV = "env";
+    private static final int ACCEPT_TIMEOUT = 10 * 1000;
+
+    public AdvancedLaunchingConnector(VirtualMachineManagerImpl virtualMachineManager) {
+        super(virtualMachineManager);
+    }
+
+    @Override
+    public Map<String, Argument> defaultArguments() {
+        Map<String, Argument> defaultArgs = super.defaultArguments();
+
+        Argument cwdArg = new AdvancedStringArgumentImpl(CWD, "Current working directory", CWD, false);
+        cwdArg.setValue(null);
+        defaultArgs.put(CWD, cwdArg);
+
+        Argument envArg = new AdvancedStringArgumentImpl(ENV, "Environment variables", ENV, false);
+        envArg.setValue(null);
+        defaultArgs.put(ENV, envArg);
+
+        return defaultArgs;
+    }
+
+    @Override
+    public String name() {
+        return "com.microsoft.java.debug.AdvancedLaunchingConnector";
+    }
+
+    @Override
+    public VirtualMachine launch(Map<String, ? extends Argument> connectionArgs)
+            throws IOException, IllegalConnectorArgumentsException, VMStartException {
+        String cwd = connectionArgs.get(CWD).value();
+        File workingDir = null;
+        if (cwd != null && Files.isDirectory(Paths.get(cwd))) {
+            workingDir = new File(cwd);
+        }
+
+        String[] envVars = null;
+        try {
+            envVars = DebugUtility.decodeArrayArgument(connectionArgs.get(ENV).value());
+        } catch (IllegalArgumentException e) {
+            // do nothing.
+        }
+
+        SocketListeningConnectorImpl listenConnector = new SocketListeningConnectorImpl(
+                virtualMachineManager());
+        Map<String, Connector.Argument> args = listenConnector.defaultArguments();
+        ((Connector.IntegerArgument) args.get("timeout")).setValue(ACCEPT_TIMEOUT); //$NON-NLS-1$
+        String address = listenConnector.startListening(args);
+
+        String[] cmds = constructLaunchCommand(connectionArgs, address);
+        Process process = Runtime.getRuntime().exec(cmds, envVars, workingDir);
+
+        VirtualMachineImpl vm;
+        try {
+            vm = (VirtualMachineImpl) listenConnector.accept(args);
+        } catch (IOException | IllegalConnectorArgumentsException e) {
+            process.destroy();
+            throw new VMStartException(String.format("VM did not connect within given time: %d ms", ACCEPT_TIMEOUT), process);
+        }
+
+        vm.setLaunchedProcess(process);
+        return vm;
+    }
+
+    private static String[] constructLaunchCommand(Map<String, ? extends Argument> launchingOptions, String address) {
```

Comment:
Can we use the implementation from the base class?

Reply:
fixed.